sequencer/src/block/full_payload/payload.rs

```diff
@@ -0,0 +1,284 @@
+        let ns_payload_range = self.ns_table().ns_range(index, &self.byte_len());
+        self.read_ns_payload(&ns_payload_range)
+    }
+
+    pub(in crate::block) fn byte_len(&self) -> PayloadByteLen {
+        PayloadByteLen(self.ns_payloads.len())
+    }
+
+    // PRIVATE HELPERS START HERE
+
+    /// Need a sync version of [`BlockPayload::from_transactions`] in order to impl [`BlockPayload::empty`].
+    fn from_transactions_sync(
+        transactions: impl IntoIterator<Item = <Self as BlockPayload<SeqTypes>>::Transaction> + Send,
+        _validated_state: &<Self as BlockPayload<SeqTypes>>::ValidatedState,
+        instance_state: &<Self as BlockPayload<SeqTypes>>::Instance,
+    ) -> Result<
+        (Self, <Self as BlockPayload<SeqTypes>>::Metadata),
+        <Self as BlockPayload<SeqTypes>>::Error,
+    > {
+        // accounting for block byte length limit
+        let max_block_byte_len: usize = u64::from(instance_state.chain_config.max_block_size)
+            .try_into()
+            .map_err(|_| <Self as BlockPayload<SeqTypes>>::Error::BlockBuilding)?;
+        let mut block_byte_len = NsTableBuilder::fixed_overhead_byte_len();
+
+        // add each tx to its namespace
+        let mut ns_builders = HashMap::<NamespaceId, NsPayloadBuilder>::new();
+        for tx in transactions.into_iter() {
+            // accounting for block byte length limit
+            block_byte_len += tx.payload().len() + NsPayloadBuilder::tx_overhead_byte_len();
+            if !ns_builders.contains_key(&tx.namespace()) {
+                // each new namespace adds overhead
+                block_byte_len += NsTableBuilder::ns_overhead_byte_len()
+                    + NsPayloadBuilder::fixed_overhead_byte_len();
+            }
+            if block_byte_len > max_block_byte_len {
+                tracing::warn!("transactions truncated to fit in maximum block byte length {max_block_byte_len}");
+                break;
+            }
+
+            let ns_builder = ns_builders.entry(tx.namespace()).or_default();
+            ns_builder.append_tx(tx);
+        }
+
+        // build block payload and namespace table
+        let mut payload = Vec::new();
+        let mut ns_table_builder = NsTableBuilder::new();
+        for (ns_id, ns_builder) in ns_builders {
+            payload.extend(ns_builder.into_bytes());
+            ns_table_builder.append_entry(ns_id, payload.len());
+        }
+        let ns_table = ns_table_builder.into_ns_table();
+        let metadata = ns_table.clone();
+        Ok((
+            Self {
+                ns_payloads: payload,
+                ns_table,
+            },
+            metadata,
+        ))
+    }
+}
+
+#[async_trait]
+impl BlockPayload<SeqTypes> for Payload {
+    type Error = crate::Error;
+    type Transaction = Transaction;
+    type Instance = NodeState;
+    type Metadata = NsTable;
+    type ValidatedState = ValidatedState;
+
+    // TODO(BlockPayload): return type should not include `Self::Metadata`
```

Comment:
I think for TODOs referring to other repos, it is better to create issues in the appropriate repos and assign them to the PL

Reply:
Posted new issues and linked to them in https://github.com/EspressoSystems/espresso-sequencer/pull/1499/commits/79308305b6b9d49015e44c685bb3289d3feeba4f

New issues spawned from this PR:

# This repo

- #1577 
  - #1574
  - #1575 
  - #1576 

# Hotshot repo

- https://github.com/EspressoSystems/HotShot/issues/3298
- https://github.com/EspressoSystems/HotShot/issues/3300

# hotshot-query-service repo

- https://github.com/EspressoSystems/hotshot-query-service/issues/639